src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ContextAware.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using System.Runtime.Serialization.DataContracts;
+
+namespace System.Runtime.Serialization
+{
+    internal sealed class ContextAwareDataContractIndex
+    {
+        private (DataContract? strong, WeakReference<DataContract>? weak)[] _contracts;
+        private ConditionalWeakTable<Type, DataContract> _keepAlive;
+
+        public int Length => _contracts.Length;
+
+        public ContextAwareDataContractIndex(int size)
+        {
+            _contracts = new (DataContract?, WeakReference<DataContract>?)[size];
+            _keepAlive = new ConditionalWeakTable<Type, DataContract>();
+        }
+
+        public DataContract? GetItem(int index) => _contracts[index].strong ?? (_contracts[index].weak?.TryGetTarget(out DataContract? ret) == true ? ret : null);
+
```

Comment:
It's only a style thing, but did you consider having providing an indexer property instead of switching to Get and Set methods?

Reply:
I found [the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-7.0#thread-safety) that says it is thread safe.

Reply:
Sort of an evolution of the context-aware side of the design that got away from the original usage pattern. An indexer would fit the original code. But the nullable notation of the original code was incorrect, even if our code was dilligent about null-checking. Maybe I've still got Monday brain, but I don't know of a way to make the getter nullable and the setter non-nullable... notation-wise anyway.